components/eamxx/src/physics/p3/eamxx_p3_process_interface.hpp

```diff
@@ -259,12 +268,16 @@ class P3Microphysics : public AtmosphereProcess
         diag_eff_radius_qc(icol,ipack) *= 1e6;
         diag_eff_radius_qi(icol,ipack) *= 1e6;
         diag_eff_radius_qr(icol,ipack) *= 1e6;
-      } // for ipack
+      }); // for ipack
 
       // Microphysics can be subcycled together during a single physics timestep,
-      // therefore we must accumulate these fluxes
-      precip_liq_surf_mass(icol) += precip_liq_surf_flux(icol) * PC::RHO_H2O * m_dt;
-      precip_ice_surf_mass(icol) += precip_ice_surf_flux(icol) * PC::RHO_H2O * m_dt;
+      // therefore we must accumulate these fluxes.
+      // Note: we need to ensure that only a single thread within the team is
+      //       updating the mass value.
+      Kokkos::single(Kokkos::PerTeam(team), [&] {
+        precip_liq_surf_mass(icol) += precip_liq_surf_flux(icol) * PC::RHO_H2O * m_dt;
```

Comment:
I wonder if we should rm the precip handling from p3_postproc, and just issue something like this at the end of run_impl:
```c++
auto precip_liq_mass = get_field_out("precip_liq_surf_mass");
auto precip_liq_flux = get_field_out("precip_liq_surf_flux");
precip_liq_mas.update(precip_liq_flux,PC::RHO_H20*dt);
```
The `Field::update` call will fully utilize the GPU, rather than just 1 thread per team.

Reply:
Ah, wait, the flux view is from the buffer, it's not a field. We do have a constructor for `Field` that uses a pre-existing dev view though... So one could do (during init)
```c++
auto layout = get_field_out("precip_liq_surf_mass").get_header().get_identifier().get_layout();
FieldIdentifier liq_flux_id ("precip_liq_surf_mass",layout,m/s,liq_mass_id.get_grid_name());
m_precip_liq_flux = Field(liq_flux_id,m_buffer.precip_liq_surf_flux);
```

Not sure if it's worth doing though...

Reply:
I'll probably leave it as is.

Would it be useful to have functions like `Field::update` be able to take a view of the correct layout/dims?